include/oneapi/dpl/internal/reduce_by_segment_impl.h

```diff
@@ -138,6 +138,293 @@ reduce_by_segment_impl(Policy&& policy, InputIterator1 first1, InputIterator1 la
 }
 
 #if _ONEDPL_BACKEND_SYCL
+
+template <typename _ExecutionPolicy, typename _Range1, typename _Range2, typename _Range3, typename _Range4,
+          typename _BinaryPredicate, typename _BinaryOperator>
+oneapi::dpl::__internal::__enable_if_hetero_execution_policy<_ExecutionPolicy,
+                                                             oneapi::dpl::__internal::__difference_t<_Range3>>
+sycl_reduce_by_segment(_ExecutionPolicy&& __exec, _Range1&& __keys, _Range2&& __values, _Range3&& __out_keys,
+                       _Range4&& __out_values, _BinaryPredicate __binary_pred, _BinaryOperator __binary_op)
+{
+    using __diff_type = oneapi::dpl::__internal::__difference_t<_Range1>;
+    using __key_type = oneapi::dpl::__internal::__value_t<_Range1>;
+    using __val_type = oneapi::dpl::__internal::__value_t<_Range2>;
+    using __flag_type = bool;
+
+    const auto __n = __keys.size();
+
+    constexpr int __vals_per_item = 2; // Each work item serially processes 2 items. Best observed performance on gpu
```

Comment:
Same comment as for scan.  Do we want to hardcode this? Does this need to be selected based on the device, or do we need to do a more general examination of a best global value here?